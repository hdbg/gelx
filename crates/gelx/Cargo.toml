[package]
name = "gelx"
version = { workspace = true }
authors = { workspace = true }
categories = ["database"]
edition = { workspace = true }
homepage = { workspace = true }
keywords = ["gel", "database", "typesafe", "macros", "codegen"]
license = { workspace = true }
readme = "readme.md"
repository = { workspace = true }
rust-version = { workspace = true }
description = "Generate fully typed rust code from your gel schema and inline queries."

[package.metadata.docs.rs]
all-features = true
rustdoc-args = ["--cfg", "docsrs"]

[dependencies]
async-trait = { workspace = true, default-features = true, optional = true }
bigdecimal = { workspace = true, default-features = true, optional = true }
bytes = { workspace = true, default-features = true }
cfg-if = { workspace = true, default-features = true }
chrono = { workspace = true, default-features = true, optional = true }
derive_more = { workspace = true, default-features = true, features = ["from", "into", "deref", "deref_mut"] }
document-features = { workspace = true, default-features = true }
gel-derive = { workspace = true, default-features = true, optional = true }
gel-errors = { workspace = true, default-features = true }
gel-protocol = { workspace = true, default-features = true }
gel-tokio = { workspace = true, default-features = true, features = ["unstable", "derive"], optional = true }
gelx_macros = { workspace = true, default-features = true }
geo = { workspace = true, default-features = true, optional = true }
geo-traits = { workspace = true, default-features = true, optional = true }
geo-types = { workspace = true, default-features = true, optional = true }
num-bigint = { workspace = true, default-features = true, optional = true }
num-traits = { workspace = true, default-features = true, optional = true }
serde = { workspace = true, default-features = true, features = ["derive"], optional = true }
serde_bytes = { workspace = true, default-features = true, optional = true }
strum = { workspace = true, default-features = true, features = ["derive"], optional = true }
typed-builder = { workspace = true, default-features = true, optional = true }
uuid = { workspace = true, default-features = true }
wkb = { workspace = true, default-features = true, optional = true }

[dev-dependencies]
assert2 = { workspace = true, default-features = true }
gelx_core = { workspace = true, default-features = true }
insta = { workspace = true, default-features = true, features = ["ron", "json", "redactions", "filters"] }
proc-macro2 = { workspace = true, default-features = true }
rstest = { workspace = true, default-features = true }
rustversion = { workspace = true, default-features = true }
test-log = { workspace = true, default-features = true, features = ["log", "trace"] }
tokio = { workspace = true, default-features = true, features = ["time", "test-util", "fs"] }
trybuild = { workspace = true, default-features = true }

[features]
# ! #### Default

## The default feature is `with_all`.
default = [
	"with_all",
	"builder",
	"strum",
]

# ! #### Types

## Include the `num-bigint` dependency.
with_bigint = [
	"dep:num-bigint",
	"dep:num-traits",
	"gel-protocol/with-num-bigint",
]

## Use the `bigdecimal` crate.
with_bigdecimal = [
	"dep:bigdecimal",
	"dep:num-bigint",
	"dep:num-traits",
	"gel-protocol/with-bigdecimal",
]

## Use the `chrono` crate for all dates.
with_chrono = [
	"dep:chrono",
	"gel-protocol/with-chrono",
]

## Use the `geo-types` crate for all geo types.
with_geo = [
	"dep:geo-traits",
	"dep:geo-types",
	"dep:geo",
	"dep:wkb",
]

## Include all additional types. This is included by default. Use `default-features = false` to disable.
with_all = [
	"with_bigint",
	"with_bigdecimal",
	"with_chrono",
	"with_geo",
	"gel-protocol/all-types",
]

# ! #### Behavior

## Use the `typed-builder` crate to generate the builders for the generated `Input` structs.
builder = [
	"dep:typed-builder",
	"gelx_macros/builder",
]

## Turn on the `query` and `transaction` methods and anything that relies on `gel-tokio`.
## The reason to separate this feature is to enable usage of this macro in browser environments
## where `gel-tokio` is not feasible.
##
## Adding this feature requires adding the `gel-protocol` dependency to your project.
## This is because the `Queryable` derive macro requires the `gel-protocol` crate to be present
## in the consuming crate.
query = [
	"gelx_macros/query",
	"dep:async-trait",
	"dep:gel-tokio",
	"dep:gel-derive",
]

## Enable serde for the generated code.
serde = [
	"bigdecimal?/serde-json",
	"chrono/serde",
	"dep:serde_bytes",
	"dep:serde",
	"gel-protocol/with-serde",
	"gelx_macros/serde",
	"geo-types?/serde",
	"geo?/use-serde",
	"num-bigint?/serde",
	"uuid/serde",
]

## Enable strum for the generated code.
##
## Adding this feature requires adding the `strum` dependency to your project.
strum = [
	"gelx_macros/strum",
	"dep:strum",
]

[lints]
workspace = true
