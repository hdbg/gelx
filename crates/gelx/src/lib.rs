#![doc = include_str!(concat!(env!("CARGO_MANIFEST_DIR"), "/cargo.md"))]
//! ## Features
#![doc = document_features::document_features!()]
#![cfg_attr(docsrs, feature(doc_cfg))]

/// Generates a query module from a query string.
///
/// ```rust
/// use gelx::gelx;
///
/// gelx!(get_users, "select User {**}");
/// ```
///
/// This macro can be called with one argument if in the root of your crate you
/// host a folder named `queries`.
///
/// ```rust
/// use gelx::gelx;
///
/// gelx!(insert_user);
/// ```
///
/// The above code will find the file `<CRATE_ROOT>/queries/insert_user.edgeql`
/// and run the query from there.
#[macro_export]
macro_rules! gelx {
	($module:ident, $query:literal) => {
		$crate::exports::gelx_macros::gelx_raw!($module, query: $query);
	};
	($module: ident) => {
		$crate::exports::gelx_macros::gelx_raw!($module);
	};
}

/// Generates a query module from a query string relative to the root of the
/// crate this is defined in. This is useful for queries that are not placed in
/// the `queries` folder at the root of the crate.
///
/// ```rust
/// use gelx::gelx_file;
///
/// gelx_file!(insert_user, "queries/insert_user.edgeql");
/// ```
///
/// The above code can actually be replaced with the
/// `gelx!(insert_user)` macro since the file is placed in the `queries`
/// folder.
#[macro_export]
macro_rules! gelx_file {
	($module:ident, $path:literal) => {
		$crate::exports::gelx_macros::gelx_raw!($module, file: $path);
	};
}

pub mod exports {
	#[cfg(feature = "with_bigdecimal")]
	#[cfg_attr(docsrs, doc(cfg(feature = "with_bigdecimal")))]
	pub use bigdecimal;
	pub use bytes;
	#[cfg(feature = "with_chrono")]
	#[cfg_attr(docsrs, doc(cfg(feature = "with_chrono")))]
	pub use chrono;
	#[cfg(feature = "query")]
	#[cfg_attr(docsrs, doc(cfg(feature = "query")))]
	pub use gel_derive;
	pub use gel_errors;
	pub use gel_protocol;
	#[cfg(feature = "query")]
	#[cfg_attr(docsrs, doc(cfg(feature = "query")))]
	pub use gel_tokio;
	pub use gelx_macros;
	#[cfg(any(feature = "with_bigdecimal", feature = "with_bigint"))]
	#[cfg_attr(
		docsrs,
		doc(cfg(any(feature = "with_bigdecimal", feature = "with_bigint")))
	)]
	pub use num_bigint;
	#[cfg(any(feature = "with_bigdecimal", feature = "with_bigint"))]
	#[cfg_attr(
		docsrs,
		doc(cfg(any(feature = "with_bigdecimal", feature = "with_bigint")))
	)]
	pub use num_traits;
	#[cfg(feature = "serde")]
	#[cfg_attr(docsrs, doc(cfg(feature = "serde")))]
	pub use serde;
	#[cfg(feature = "builder")]
	#[cfg_attr(docsrs, doc(cfg(feature = "builder")))]
	pub use typed_builder;
	pub use uuid;
}
