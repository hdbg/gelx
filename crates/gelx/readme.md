<p align="center">
  <a href="#">
    <img width="300" src="https://raw.githubusercontent.com/ifiokjr/gelx/main/setup/assets/logo.svg"  />
  </a>
</p>

<br />

> Generate fully typed rust code from your gel schema and inline queries with `gelx`.

<br />

[![Crate][crate-image]][crate-link] [![Docs][docs-image]][docs-link] [![Status][ci-status-image]][ci-status-link] [![Unlicense][unlicense-image]][unlicense-link] [![codecov][codecov-image]][codecov-link]

## Installation

To install the `gelx` crate you can use the following commands.

```bash
cargo add gelx

# Optional dependencies
cargo add gel-protocol strum
```

Or directly add the following to your `Cargo.toml` file.

```toml
[dependencies]
gelx = "0.4"

# Optional dependencies
gel-protocol = "0.8" # needed for query feature
strum = "0.27" # needed for strum feature
```

Make sure you've [installed](https://docs.geldata.com/reference/using/cli#ref-cli-gel-install) the `gel` CLI for your platform.

Once installed you should be able to run the following command to verify your installation.

```bash
gel --version
```

And then initialize in the project directory.

```bash
gel init
```

## `gelx!`

Working with the default `gel` requires writing queries and creating the expected types for both the input and output into these queries. The correctness of your code is only checked at runtime increasing the risk of bugs and errors.

`gelx!` transforms your queries into types and functions, providing safety and peace of mind during the development of your project.

### Inline Queries

```rust
use gel_errors::Error;
use gel_tokio::create_client;
use gelx::gelx;

// Creates a module called `simple` with a function called `query` and structs
// for the `Input` and `Output`.
gelx!(
	example,
	"select { hello := \"world\", custom := <str>$custom }"
);

#[tokio::main]
async fn main() -> Result<(), Error> {
	let client = create_client().await?;
	let input = example::Input {
		custom: String::from("custom"),
	};

	// For queries the following code can be used.
	let output = example::query(&client, &input).await?;

	Ok(())
}
```

The macro above generates the following code in the background::

```rust
pub mod example {
	use ::gelx::exports as __g;
	/// Execute the desired query.
	pub async fn query(
		client: &__g::gel_tokio::Client,
		props: &Input,
	) -> ::core::result::Result<Output, __g::gel_errors::Error> {
		client.query_required_single(QUERY, props).await
	}
	/// Compose the query as part of a larger transaction.
	pub async fn transaction(
		conn: &mut __g::gel_tokio::Transaction,
		props: &Input,
	) -> ::core::result::Result<Output, __g::gel_errors::Error> {
		conn.query_required_single(QUERY, props).await
	}
	#[derive(
		Clone,
		Debug,
		__g::serde::Serialize,
		__g::serde::Deserialize,
		__g::typed_builder::TypedBuilder,
		__g::gel_derive::Queryable,
	)]
	pub struct Input {
		#[builder(setter(into))]
		pub custom: String,
	}
	impl __g::gel_protocol::query_arg::QueryArgs for Input {
		fn encode(
			&self,
			encoder: &mut __g::gel_protocol::query_arg::Encoder,
		) -> core::result::Result<(), __g::gel_errors::Error> {
			let map = __g::gel_protocol::named_args! {
				"custom" => self.custom.clone(),
			};
			map.encode(encoder)
		}
	}
	#[derive(
		Clone, Debug, __g::serde::Serialize, __g::serde::Deserialize, __g::gel_derive::Queryable,
	)]
	pub struct Output {
		pub hello: String,
		pub custom: String,
	}
	/// The original query string provided to the macro. Can be reused in your
	/// codebase.
	pub const QUERY: &str = "select { hello := \"world\", custom := <str>$custom }";
}
```

### Query Files

Define a query file in the `queries` directory of your crate called `select_user.edgeql`.

```edgeql
# queries/select_user.edgeql

select User {
  name,
  bio,
  slug,
} filter .slug = <str>$slug;
```

Then use the `gelx` macro to import the query.

```rust
use gel_errors::Error;
use gelx::create_client;
use gelx::gelx;

// Creates a module called `select_user` with public functions `transaction` and
// `query` as well as structs for the `Input` and `Output`.
gelx!(select_user);

#[tokio::main]
async fn main() -> Result<(), Error> {
	let client = create_client().await?;

	// Generated code can be run inside a transaction.
	let result = client
		.transaction(|mut txn| {
			async move {
				let input = select_user::Input {
					slug: String::from("test"),
				};
				let output = select_user::transaction(&mut txn, &input).await?;
				Ok(output)
			}
		})
		.await?;

	Ok(())
}
```

## CLI

The `gelx_cli` crate exposes a binary called `gelx` which can be used to generate the typed code into rust files rather than inline queries.

It should be run from the crate directory and will read from the configuration specified in the next section.

```bash
gelx generate --cwd path/to/crate
```

Sometimes you will need to check that the generated code matches the current database schema and queries generated by gel. This is useful for CI pipelines to ensure the generated code is up to date.

```bash
gelx check --cwd path/to/crate
```

If there are changes that haven't been accounted for, the check will fail with a diff and you should regenerate the code.

More information can be found in the [`gelx_cli` readme](https://github.com/ifiokjr/gelx/blob/main/crates/gelx_cli/readme.md).

## Configuration

The following configuration options are supported and the provided defaults will be used if not specified.

```toml
[package.metadata.gelx]
## The path to the directory containing the queries.
queries_path = "./queries"

## The features to enable and their aliases. By default all features are enabled.
## To disable a feature set it to false. To alias a feature behind a feature flag
## use the following format `feature = { query = "ssr" }`. This will enable the query
## feature only when the `ssr` feature is enabled.
##
## The available features are:
##
## - `query` - When enabled you must include `gel-protocol` as a dependency.
## - `serde` - Enable serde for the generated code.
## - `strum` - When enabled you must include `strum` as a dependency.
## - `builder` - Use the `typed-builder` crate to generate the builders for the generated `Input` structs.
features = { query = true, strum = true, builder = true,  }

## The location of the generated code when using the `gelx` cli.
output_file = "./src/gelx_generated.rs"

## The name of the arguments input struct. Will be transformed to PascalCase.
input_struct_name = "Input"

## The name of the exported output struct for generated queries. Will be transformed to PascalCase.
output_struct_name = "Output"

## The name of the query function exported.
query_function_name = "query"

## The name of the transaction function exported.
transaction_function_name = "transaction"

## The name of the query constant exported.
query_constant_name = "QUERY"

## The alias used for the `gelx::exports` module.
exports_alias = "__g"

## The macros which are always derived for the generated structs.
struct_derive_macros = ["Debug", "Clone"]

## The macros which are always derived for the generated enums.
enum_derive_macros = ["Debug", "Clone", "Copy"]

## The relative path to the `gel` config file. This is optional and if not provided the `gel`
## config will be read from the environment variables.
# gel_config_path = "./gel.toml"

## The name of the `gel` instance to use. This is optional and if not provided the environment
## variable `$GEL_INSTANCE` will be used.
# gel_instance = "$GEL_INSTANCE"

## The name of the `gel` branch to use. This is optional and if not provided the environment
## variable `$GEL_BRANCH` will be used.
# gel_branch = "$GEL_BRANCH"
```

## Future Work ⚠️

This crate is still in early development and there are several features that are not yet implemented.

### Missing Types

Currently the following types are not supported:

- `MultiRange` - The macro will panic if a multirange is used.
- `POSTGIS` - No elegant way to support this yet with integration with the `geo` crate.

#### `MultiRange`

These are not currently exported by the `gel-protocol` so should be added in a PR to the `gel-protocol` crate, if they are still supported in the new protocol.

#### `POSTGIS`

No elegant way to support this yet with integration with the `geo` crate.

Export wrapper types from the `gelx` crate which support the `POSTGIS` types.

### LSP parsing

Currently the macro depends on having a running gel instance to parse the query string.

Once an LSP is created for gel it would make sense to switch from using string to using inline gel queries.

```rust,ignore
use gelx::gelx;

gelx!(
	example,
	select User {**}
);
```

[crate-image]: https://img.shields.io/crates/v/gelx.svg
[crate-link]: https://crates.io/crates/gelx
[docs-image]: https://docs.rs/gelx/badge.svg
[docs-link]: https://docs.rs/gelx/
[ci-status-image]: https://github.com/ifiokjr/gelx/workflows/ci/badge.svg
[ci-status-link]: https://github.com/ifiokjr/gelx/actions?query=workflow:ci
[unlicense-image]: https://img.shields.io/badge/license-Unlicense-blue.svg
[unlicense-link]: https://opensource.org/license/unlicense
[codecov-image]: https://codecov.io/github/ifiokjr/gelx/graph/badge.svg?token=87K799Q78I
[codecov-link]: https://codecov.io/github/ifiokjr/gelx

## Features

- **`default`** — The default feature is `with_all`.
- **`with_bigint`** — Include the `num-bigint` dependency.
- **`with_bigdecimal`** — Use the `bigdecimal` crate.
- **`with_chrono`** — Use the `chrono` crate for all dates.
- **`with_all`** _(enabled by default)_ — Include all additional types. This is included by default. Use `default-features = false` to disable.
- **`builder`** — Use the `typed-builder` crate to generate the builders for the generated `Input` structs.
- **`query`** — Turn on the `query` and `transaction` methods and anything that relies on `gel-tokio`. The reason to separate this feature is to enable usage of this macro in browser environments where `gel-tokio` is not feasible.
- **`serde`** — Enable serde for the generated code.
- **`strum`** - Use the `strum` crate for deriving strings from the created enums.

## Contributing

[`devenv`](https://devenv.sh/) is used to provide a reproducible development environment for this project. Follow the [getting started instructions](https://devenv.sh/getting-started/).

To automatically load the environment you should [install direnv](https://devenv.sh/automatic-shell-activation/) and then load the `direnv`.

```bash
direnv allow .
```

You now have a shell with all the dependencies installed and project specific commands available.

Run the following commands to install all the required dependencies.

```bash
install:all
```

This installs all the cargo binaries locally so you don't need to worry about polluting your global namespace.

At this point you must setup the gel instance.

```bash
db:setup # setup the gel instance
```

Now you can make your changes and run tests.

```bash
test:all
```

### Available Commands

You can view all the available scripts, packages, tasks and environment variables by running `devenv info`.

- `build:all`: Build all crates with all features activated.
- `build:docs`: Build documentation site.
- `coverage:all`: Test all files and generate a coverage report for upload to codecov.
- `db:destroy`: Destroy the local database.
- `db:setup`: Setup the local database.
- `db:up`: Watch changes to the local database.
- `fix:all`: Fix all fixable lint issues.
- `fix:clippy`: Fix fixable lint issues raised by rust clippy.
- `fix:format`: Fix formatting for entire project.
- `fix:gelx`: Fix fixable lint issues raised by gelx.
- `gelx`: Install all dependencies.
- `install:all`: Install all dependencies.
- `install:cargo:bin`: Install cargo binaries locally.
- `lint:all`: Lint all project files.
- `lint:clippy`: Check rust clippy lints.
- `lint:format`: Check all formatting is correct.
- `lint:gelx`: Check gelx is formatted correctly.
- `setup:ci`: Setup the github ci environment.
- `setup:helix`: Setup the helix editor for development.
- `setup:vscode`: Setup the vscode editor for development.
- `test:all`: Test all project files.
- `update:deps`: Update all project dependencies.

### Upgrading `devenv`

If you have an outdated version of `devenv` you can update it by running the following commands. If you have an easier way, please create a PR and I'll update these docs.

```bash
nix profile list # find the index of the devenv package
nix profile remove <index>
nix profile install ---accept-flake-config nixpkgs#devenv
```

### Editor Setup

To setup recommended configuration for your favorite editor run the following commands.

```bash
setup:vscode # Setup vscode with recommended configuration
```

```bash
setup:helix # Setup helix with recommended configuration
```

## License

Unlicense, see the [license](https://github.com/ifiokjr/gelx/blob/main/license) file.
