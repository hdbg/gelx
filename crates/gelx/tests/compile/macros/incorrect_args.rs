use gelx::gelx;

fn main() {
	gelx!(example_module, invalid);
}
