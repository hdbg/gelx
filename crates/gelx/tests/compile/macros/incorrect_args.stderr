error: no rules expected `invalid`
 --> tests/compile/macros/incorrect_args.rs:4:24
  |
4 |     gelx!(example_module, invalid);
  |                           ^^^^^^^ no rules expected this token in macro call
  |
note: while trying to match meta-variable `$query:literal`
 --> src/lib.rs
  |
  |     ($module:ident, $query:literal) => {
  |                     ^^^^^^^^^^^^^^
